crmprtd/insert.py

```diff
@@ -142,85 +187,157 @@ def bisect_insert_strategy(sesh, obs):
     but in the optimal case it reduces the transactions to a constant
     1.
     """
-    log.debug("Begin mass observation insertion", extra={"num_obs": len(obs)})
+    log.debug("Begin mass observation insertion", extra={"num_obs": len(observations)})
 
     # Base cases
-    if len(obs) < 1:
+    if len(observations) < 1:
         return DBMetrics(0, 0, 0)
-    elif len(obs) == 1:
-        try:
-            # Create a nested SAVEPOINT context manager to rollback to in the
-            # event of unique constraint errors
-            log.debug("New SAVEPOINT for single observation")
-            with sesh.begin_nested():
-                sesh.add(obs[0])
-        except IntegrityError as e:
-            log.debug(
-                "Failure, observation already exists",
-                extra={"obs": obs, "exception": e},
-            )
-            db_metrics = DBMetrics(0, 1, 0)
-        except InsertionError as e:
-            log.warning(
-                "Failure occured during insertion", extra={"obs": obs, "exception": e}
-            )
-            db_metrics = DBMetrics(0, 0, 1)
-        else:
-            log.debug("Success for single observation")
-            log.info("Successfully inserted observations: 1")
-            db_metrics = DBMetrics(1, 0, 0)
-        sesh.commit()
-        return db_metrics
-
+    elif len(observations) == 1:
+        return insert_single_obs(sesh, observations[0])
     # The happy case: add everything at once
     else:
         try:
             with sesh.begin_nested():
-                log.debug("New SAVEPOINT", extra={"num_obs": len(obs)})
-                sesh.add_all(obs)
+                log.debug("New SAVEPOINT", extra={"num_obs": len(observations)})
+                sesh.add_all(observations)
         except IntegrityError:
             log.debug("Failed, splitting observations.")
             sesh.rollback()
 
-            a, b = split(obs)
+            a, b = split(observations)
             dbm_a = bisect_insert_strategy(sesh, a)
             dbm_b = bisect_insert_strategy(sesh, b)
             return dbm_a + dbm_b
         else:
             log.info(
-                f"Successfully inserted observations: {len(obs)}",
-                extra={"num_obs": len(obs)},
+                f"Successfully inserted observations: {len(observations)}",
+                extra={"num_obs": len(observations)},
             )
-            db_metrics = DBMetrics(len(obs), 0, 0)
+            db_metrics = DBMetrics(len(observations), 0, 0)
         sesh.commit()
         return db_metrics
 
 
-def insert(sesh, observations, sample_size):
+def chunk_bisect_insert_strategy(sesh, observations):
+    log.info("Using Chunk + Bisection Strategy")
+    dbm = DBMetrics(0, 0, 0)
+    for chunk in bisection_chunks(observations):
+        dbm += bisect_insert_strategy(sesh, chunk)
+    return dbm
+
+
+def obs_to_pg_insert_dict(obs):
```

Comment:
Much better.

Reply:
Stupid thing is that I knew I should do this all along....